kata deploy rewrites

egernst opened the pull request:
Simplify the yaml and combine the prior scripts. The resulting script,
kata-deploy.sh, is used for  install and configuration and
removal for CRI-O and containerd. While this could be used standalone
outside of daemonsets, today it will sleep infinity after processing the
request, since it is assumed to be called by a daemon.

By checking the CRI runtime within the script itself, we no longer need
to support many daemonsets for deploy - just a single. Still requires a
seperate cleanup daemonset (for restarting the CRI runtime), and an
RBAC.

Verified with CRI-O -- containerd testing WIP

Throwing this up now for feedback since I do not bash good.

Todo:
- [x] test with docker
- [x] test with crio
- [x] update readme

egernst commented on kata-deploy/Dockerfile:
```diff
@@ -12,7 +11,7 @@ tar xvf ${KATA_FILE} -C /opt/kata-artifacts/ && \
 
 RUN \
-curl -s -o /bin/kubectl https://storage.googleapis.com/kubernetes-release/release/${KUBECTL_VER}/bin/linux/amd64/kubectl && \
+curl -Lso /bin/kubectl https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && \
```
Ah.  Either way - let's make that a variable which can be override with --build-arg.  Good call.

egernst commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,144 @@
+
+function install_artifacts() {
+	echo "copying kata artifacts onto host"
+	cp -R /opt/kata-artifacts/opt/kata/* /opt/kata/
```
done

jodh-intel commented:
Thanks @egernst.

lgtm

jodh-intel commented:
@egernst - can we remove the dnm label now?

@kata-containers/packaging  - yo! PR to remove guys! :)

mcastelino commented on kata-deploy/scripts/kata-deploy-docker.sh:
```diff
@@ -0,0 +1,123 @@
+{
+  "runtimes": {
+    "kata-qemu": {
+      "path": "/opt/kata/bin/kata-qemu"
```
@egernst do we need to do this. The default kata with the associated toml will always use QEMU. Using this extra step will slow down the launch. Maybe easier to just soft link the qemu toml as the default toml for efficiency,

egernst replied:
/me wonders how much this'll actually slow it down?

egernst replied:
Ah, I see - the toml would be an interesting way -- I'm just not sure its worth the inconsistency between the two hyerpvisors.  This makes it more clear/explicit.  Let's try to measure the delta?

mcastelino replied:
Another way is to just have a kata-runtime in addition. In any case I am OK either way.

egernst replied:
I’m not sure this is available for crio runtimeClass. Let’s take a look as a future improvement? Could save exec time.

bergwolf replied:
But this is the docker config file and it is supported by docker.

egernst replied:
Right, but in 1.5, we have the wrapper already available (and shared as part of the dockerfuke between docker/cri). Let me play with this though; I’m curious how much the wrapper costs.

egernst replied:
fixed for docker!

egernst commented:
Tested with CRIO and docker.  Not able to get runtimeClass working for containerd yet, but I'd prefer to get this enabled in a future PR.

@jodh-intel I addressed your comments - can you PTAL?

jodh-intel commented on kata-deploy/scripts/kata-deploy-docker.sh:
```diff
@@ -0,0 +1,110 @@
+		cp -n "$conf_file" "$conf_file_backup"
+
+		# Merge in the json snippet:
+		jq -s '[.[] | to_entries] | flatten | reduce .[] as $dot ({}; .[$dot.key] += $dot.value)' "${conf_file}" "${snippet}" > "${tmp_file}"
```
Impressive!

/me wonders idly if IOJQCC is a thing... ;)

jodh-intel commented:
Looks like pullapprove is "stuck" so re-applying/removing the dnm label to see if that clears it...

jodh-intel commented:
nada. Well, pullapprove will cease to be a concern in a couple of days, so we can just handle this PR "manually".

Could you tal @grahamwhaley ?

grahamwhaley commented:
@jodh-intel - it might be as there is `RFC` in the subject line....

jodh-intel commented:
@grahamwhaley - doh! you're right! I've removed that prefix @egernst, so please can folk review so we can finally get it landed...

jodh-intel commented:
/test

jodh-intel commented:
Ping @kata-containers/packaging - one more ack required over here y'all!

raravena80 commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,144 @@
+		echo invalid-node name
+		exit
+	fi
+	echo "$runtime" | awk -F'[:]' '/Container Runtime Version/ {print $2}' | tr -d ' '
```
kubectl has `-o=jsonpath='{...}'`, that may be handy.

raravena80 commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,178 @@
+      [plugins.cri.containerd.untrusted_workload_runtime]
+        runtime_type = "io.containerd.runtime.v1.linux"
+        runtime_engine = "/opt/kata/bin/kata-runtime"
+        runtime_root = ""
```
Might be a good idea to add the `runtimeClass` option.

egernst replied:
see: https://github.com/kata-containers/packaging/issues/323

bergwolf replied:
Also it would be good to use shimv2 instead of kata-runtime for containerd.

egernst replied:
Agreed! Getting both “v2” interfaces is the logical next step here.

egernst commented:
/test

bergwolf commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,178 @@
+}
+
+function reset_runtime() {
+	kubectl label node $NODE_NAME kata-containers.io/container-runtime- kata-containers.io/kata-runtime-
```
here `kata-containers.io/container-runtime-` is used. So in deploy.yaml we should lable node accordingly right?

gnawux replied:
hmm, our official site URL is `katacontainers.io`, while here is `kata-containers.io`...

I know this is not introduced by this PR, just wonder should we fix this? @egernst

bergwolf commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,178 @@
+function reset_runtime() {
+	kubectl label node $NODE_NAME kata-containers.io/container-runtime- kata-containers.io/kata-runtime-
+	systemctl daemon-reload
+	systemctl restart $1
```
restarting `daemon-reload`? I don't quite understand that one.

gnawux replied:
if you changed any systemd configuration, you should do a `systemctl daemon-reload`

bergwolf replied:
well, that is done one line above and followed by `systemctl restart daemon-reload`.

egernst replied:
$1 is expected to be either containerd or crio.  We're restarting these.

bergwolf replied:
oops, my bad, sorry for the noise

bergwolf commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,178 @@
+	fi
+
+	# only install / remove / update if we are dealing with CRIO or containerd
+	if [ "$runtime" == "crio" ] || [ "$runtime" == "containerd" ]; then
```
What if runtime is dockershim? Should we bail out?

egernst replied:
Nah, because the daemonset will restart then. Just go sleep if docker.

egernst replied:
Nah, because the daemonset will restart then. Just go sleep if docker.

egernst commented on kata-deploy/kata-deploy.yaml:
```diff
@@ -1,44 +1,6 @@
-apiVersion: apps/v1
-kind: DaemonSet
 metadata:
   name: kubelet-cri-o-kata
```
That’s in error. Will fix! It’ll be same for both. Delete/copy paste error on my part.

egernst replied:
done!

egernst commented on kata-deploy/scripts/kata-deploy.sh:
```diff
@@ -0,0 +1,178 @@
+
+	cat <<EOT | tee -a "$crio_conf_file"
+[crio.runtime.runtimes.kata-qemu]
+  runtime_path = "/opt/kata/bin/kata-qemu"
```
I'm not sure of this.  I'll update docker, and see what is available for CRIO for conmon as well as for the "v2" interface
